fc/parsing/actions.py

```diff
@@ -1200,204 +1268,353 @@ def _number_generator(self, value, units):
-
-                # Add the new input variable, with ontology annotation
-                variable = self._create_variable(var)
-
-            # Store initial value if given
-            if var.initial_value is not None:
-                self.initial_values[var.rdf_term] = var.initial_value
-
-            # Maintain link to time variable, if needed
-            is_time = variable is self.time_variable
-
-            # Convert units if needed
-            if var.units is not None:
-                units = self.units.get_unit(var.units)
-                if units != variable.units:
-                    variable = self.model.convert_variable(variable, units, DataDirectionFlow.INPUT)
-
-                    # Update cached time variable
-                    if is_time:
-                        self.time_variable = variable
-
-    def _add_local_variables(self):
-        """
-        Modify the model by adding any variables defined in ``var`` statements.
-        """
-        for var in self.local_var_declarations:
+                # Variable not found: Still OK, as long as it refers to an existing model variable
+                try:
+                    self._variable_generator(name)
+                except MissingVariableError:
+                    raise ProtocolError(f'Define or clamp statement found for unknown variable: {name}.')
 
-            # Variable names must be unique (can't even be re-used in imported/nested protocols)
-            if var.name in self.local_vars:
-                raise ProtocolError(f'Variable "{var.name}" was defined by more than one var statement.')
+                # Create new protocol variable to store info
+                pvar = name_to_pvar[name] = ProtocolVariable(name)
 
-            # Create and store variable
-            name = self.model.get_unique_name('protocol__' + var.name)
-            units = self.units.get_unit(var.units)
-            self.local_vars[var.name] = self.model.add_variable(name, units, initial_value=var.initial_value)
+            # Store equation
+            pvar.update(equation=eq)
 
-    def _add_define_variables(self):
-        """
-        Modify the model by adding any unknown variables appearing on the LHS of a ``define`` statement (or an
-        equivalent clamp-to statement), that are also listed as an output or optional input, and have units defined.
-        """
-        # Get all variables named by inputs and outputs
-        # Note: Don't worry about clashes here, this is checked later in self._sanity_check
-        variables = {var.prefixed_name: var for var in self.inputs + self.outputs if var.units is not None}
+            # Any model variable that is redefined by an equation is treated as an input
+            if not pvar.local:
+                pvar.update(input=True)
 
-        for eq in self.equations:
+        # Resolve references to model variables
+        # If multiple references point to the same model variable, merge them
```

Comment:
We should keep a record of the references though so they appear as distinct inputs/outputs. Though I don't think we have any instances of that in models!

Reply:
Wouldn't it be an error for two inputs to point to the same variable that way?

Reply:
It would kind-of indicate an inconsistency in the ontologies used, since a variable shouldn't is-a two different concepts in the way we've set up the oxmeta ontology. But this is possible in other ontologies since they consider things slightly differently.

Reply:
But it could create a situation where I say "set x to 1", and then "set y to 2", not realising they both set `model.z` ?

Reply:
So a warning at the very least, but I'm leaning towards an error, because a user probably doesn't want a simulation to run this way?

Reply:
That would be reasonable in the first instance.